x-pack/legacy/plugins/code/server/utils/repository_reference_helper.ts

```diff
@@ -0,0 +1,171 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SavedObject, SavedObjectsClientContract } from 'src/core/server';
+import { SAVED_OBJ_REPO_REF } from '../../common/constants';
+
+export interface RepositoryReferenceHelper {
+  /**
+   * Creates a reference from the current namespace to the given repository.
+   *
+   * @param uri The uri of the repository.
+   * @returns true if there is no other references for the repository, false otherwise.
+   */
+  createReference(uri: string): Promise<boolean>;
+
+  /**
+   * Checks whether there is a reference from the current namespace to the given repository.
+   *
+   * @param uri The uri of the repository.
+   * @returns true if there is a reference from the current namespace to the given repository, false otherwise.
+   */
+  hasReference(uri: string): Promise<boolean>;
+
+  /**
+   * Throws an error if there is no reference from the current namespace to the given repository.
+   * there is none.
+   *
+   * @param uri The uri of the repository.
+   */
+  ensureReference(uri: string): Promise<void>;
+
+  /**
+   * Deletes the reference from the current namespace to the given repository.
+   *
+   * @param uri The uri of the repository.
+   * @returns True if there is no more references to the repository after the deletion.
+   */
+  deleteReference(uri: string): Promise<boolean>;
+
+  /**
+   * Finds all repository uris of which the current namespace has references.
+   *
+   * @returns uris the current namespace references to.
+   */
+  findReferences(): Promise<string[]>;
+}
+
+/**
+ * A factory function helps to create a appropriate helper instance.
+ *
+ * @param client A saved objects client instance.
+ * @returns An helper instance.
+ */
+export function getReferenceHelper(client: SavedObjectsClientContract): RepositoryReferenceHelper {
+  return new SharedObjectReferenceHelper(client);
+}
+
+class SharedObjectReferenceHelper implements RepositoryReferenceHelper {
+  constructor(private readonly client: SavedObjectsClientContract) {}
+
+  /**
+   * TODO better way to access the current space
+   */
+  private namespace(): string {
+    // @ts-ignore
+    return this.client.spaceId;
+  }
+
+  async createReference(uri: string): Promise<boolean> {
+    let obj: SavedObject<any> | undefined;
+    try {
+      obj = await this.client.get(SAVED_OBJ_REPO_REF, uri);
+    } catch (e) {
+      obj = undefined;
+    }
+    if (obj === undefined) {
+      await this.client.create(
+        SAVED_OBJ_REPO_REF,
+        {
+          uri,
+          namespaces: [this.namespace()],
+        },
+        {
+          id: uri,
+        }
+      );
+      return true;
+    } else {
+      const namespaces = new Set((obj.attributes as RepositoryReferences).namespaces);
+      if (namespaces.has(this.namespace())) {
+        throw new Error(`Reference to [${uri}] from [${this.namespace()}] already exists`);
```

Comment:
do we need to throw an error in here? or just return true?

Reply:
Throwing the error here is to keep the behavior while trying to import an already existing repo.

Reply:
The design of the method is to inform the caller if the reference already exists, so that the caller may tell the user that "the repository has already been imported", and the return value is used to distinguish between whether the reference is the first to the repo.